Swim with the Mermaids

rogeruiz opened the pull request:
:fish:

Docs: https://github.com/18F/cg-site/blob/27302cc2abc05ae377ab7dd2a3421cb15b2542aa/README.md#add-diagrams-to-documentation-pages

rogeruiz committed:
Add swimWithTheMermaids :fish:

jmcarp commented on static_src/js/main.js:
```diff
@@ -267,7 +267,25 @@ function initializeJS() {
+    return undefined;
+  }
+
+  mermaid.mermaidAPI.render(
```
Does this assume there's either zero diagrams or one? Since we're selecting by (hopefully unique) id, would this work with multiple diagrams?

rogeruiz replied:
yeah, I think I'm going to have to pass in nodes rather than a string to the render function, for now I'm just copying over Scuttle, but I'm not sure how I'm gonna get the styles over. Probably going to need to publish it to get it over in `cg-style` as a `npm install` 🤷‍♂️

rogeruiz committed:
Bring in the styles from Scuttle;

Manually for now. Scuttle's not ready to output these styles in a way
external dependencies can easily consume them... for now.

rogeruiz committed:
Bring in new dependencies;

This commit brings in Scuttle, but doesn't use it. For now it's a little
reminder of how projects should be consuming Scuttle from the outside.

rogeruiz committed:
Bring in all the diagram assets;

- new Shortcode
- Styles
- JavaScript for initializing

jmcarp commented on package.json:
```diff
@@ -26,9 +26,11 @@
     "node-sass": "^3.10.0",
     "onchange": "git+https://git@github.com/msecret/onchange.git",
-    "postcss-import": "^8.1.2"
+    "postcss-import": "^8.1.2",
```
I see that we're importing `mermaid`, but where do we use `postcss-import` or `scuttle`?

rogeruiz replied:
so I believe, that `postcss-import` was already there. I'm hoping to publish Scuttle `1.1.0` which will include the compiled CSS in the package so we can copy it out the same way that `cg-style` is copied into the project from `node_modules`. We can remove it for now though, it's not being used at the moment.

rogeruiz replied:
in other words, I'm not really sure where we use `postcss-import` 😁

It might be good to clean things up, but I don't want to turn this into a huge amount of work to refactor my way out of this frontend paper bag.

jmcarp replied:
OK, seems like we can leave `scuttle` out until installing it actually does something.

jmcarp commented on static_src/js/main.js:
```diff
@@ -267,7 +272,37 @@ function initializeJS() {
+        $el.find( '.js-diagrams__stage' ).html( c );
+      }
+    );
+    $el.find( '.js-diagrams__stage' ).find( 'style' ).remove();
```
Could we do this with css instead of js?

rogeruiz replied:
it's coming [from the mermaid library here](https://github.com/knsv/mermaid/blob/d6e22d5f5e969d78d6b2e9ab85aaa62119a18b84/src/mermaidAPI.js#L437-L448). so we've got to remove it after the fact for now.


![](https://images.duckduckgo.com/iu/?u=http%3A%2F%2Fstatic.tvtropes.org%2Fpmwiki%2Fpub%2Fimages%2Fcampfire.png&f=1)

jmcarp commented on static_src/js/main.js:
```diff
@@ -267,7 +272,37 @@ function initializeJS() {
 jQuery(document).ready(function(){
     initializeJS();
     $('[rel=show-github]').showGithub();
+    swimWithTheMermaids();
```
Sorry to be un-fun, but wdyt about calling this something more descriptive like `renderDiagrams`?

jmcarp commented on static_src/js/.eslintrc:
```diff
@@ -0,0 +1,18 @@
+{
```
I'm all in favor of checking in some linting rules, but if we do, we should probably run `eslint` in ci.

rogeruiz replied:
ah dang, fat-fingered this commit. do you think I should remove it or add the CI tests? I'm up for either. but if I add lint file, I'll want to update this one so it's... more opinionated. 🤷‍♂️

rogeruiz committed:
Better name;

h/t @jmcarp

mogul reviewed (commented):
I think this is great, but I'm unqualified to opine as to the technical implementation.

rogeruiz committed:
Add documentation for creating diagrams

rogeruiz committed:
Update diagram to use Scuttle

rogeruiz committed:
Update content a bit

rogeruiz committed:
Remove ambitious Scuttle package;

h/t @jmcarp

rogeruiz committed:
Revert "Update content a bit"

This reverts commit 39db45cc04634746033668bc1462d2b98a165e7b.
